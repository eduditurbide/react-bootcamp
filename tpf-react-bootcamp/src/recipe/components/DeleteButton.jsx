import { SaveOutlined } from '@mui/icons-material'
import { Button, Dialog, DialogActions, DialogContent, DialogContentText, DialogTitle, useMediaQuery, useTheme } from '@mui/material'
import React, { useState } from 'react'

export const DeleteButton = ({
  title = '',
  text = '',
  onAgreeClickEvent = null
}) => {

  const [open, setOpen] = useState(false);
  const theme = useTheme();
  const fullScreen = useMediaQuery(theme.breakpoints.down('md'));

  const handleClickOpen = () => {
    setOpen(true);
  };

  const handleClose = () => {
    setOpen(false);
  };

  return (
    <>
      <Button
        // variant="contained"
        color="error"
        sx={{ padding: 1, width: { xs: '100%', sm: 'auto' } }}
        onClick={handleClickOpen}
      >
        <SaveOutlined sx={{ fontSize: 30, mr: 1 }} />
        Eliminar
      </Button>
      <Dialog
        fullScreen={fullScreen}
        open={open}
        onClose={handleClose}
        aria-labelledby="responsive-dialog-title"
      >
        <DialogTitle id="responsive-dialog-title">
          {title}
        </DialogTitle>
        <DialogContent>
          <DialogContentText>
            {text}
          </DialogContentText>
        </DialogContent>
        <DialogActions>
          <Button autoFocus onClick={handleClose}>
            Cancelar
          </Button>
          <Button onClick={e => { handleClose(); onAgreeClickEvent(e); }} autoFocus>
            Aceptar
          </Button>
        </DialogActions>
      </Dialog>
    </>
  )
}
