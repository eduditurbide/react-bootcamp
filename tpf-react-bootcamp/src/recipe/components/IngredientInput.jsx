import { useState } from 'react'
import { TextField } from '@mui/material'

export const IngredientInput = ({ name = "", value = "", onBlurEvent = null }) => {
  const [ingredientValue, setIngredientValue] = useState(value || "")
  
  return (
    <TextField
      type="text"
      variant="filled"
      fullWidth
      sx={{ border: 'none', mb: 3 }}
      placeholder="Ingrese un ingrediente"
      name={name}
      value={ingredientValue}
      onChange={({ target }) => setIngredientValue(target.value)}
      onBlur={e => onBlurEvent(e)}
    />
  )
}
