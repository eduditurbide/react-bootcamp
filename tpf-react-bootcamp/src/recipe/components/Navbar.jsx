import { Menu as MenuIcon, ExitToAppOutlined, ArrowRight } from "@mui/icons-material";
import { AppBar,  Box, Button, Chip, IconButton, ListItemIcon, Menu, MenuItem, Toolbar, Typography } from "@mui/material"
import { useContext, useState } from "react";
import { Link, useNavigate } from "react-router-dom";
import { AuthContext } from "../../auth/context";

export const Navbar = () => {

  const navigate = useNavigate();
  const { user, logout } = useContext(AuthContext);
  const [anchorElNav, setAnchorElNav] = useState(null);

  const handleOpenNavMenu = (event) => {
    setAnchorElNav(event.currentTarget);
  };

  const handleCloseNavMenu = () => {
    setAnchorElNav(null);
  };

  const onLogout = () => {
    handleCloseNavMenu()
    logout()
    navigate('/auth/login', {
      replace: true
    })
  }

  return (
    <AppBar position="fixed">
      <Toolbar>
        
        <Typography
          variant="h6"
          noWrap
          component="a"
          href="/"
          sx={{
            mr: 3,
            mb: '5px',
            display: { xs: 'none', md: 'flex' },
            fontFamily: 'monospace',
            fontWeight: 700,
            letterSpacing: '.3rem',
            color: 'inherit',
            textDecoration: 'none',
          }}
        >
          Recipes App
        </Typography>

        <Box sx={{ flexGrow: 1, display: { xs: 'flex', md: 'none' } }}>
          <IconButton
            size="large"
            aria-label="account of current user"
            aria-controls="menu-appbar"
            aria-haspopup="true"
            onClick={handleOpenNavMenu}
            color="inherit"
          >
            <MenuIcon />
          </IconButton>
          <Menu
            id="menu-appbar"
            anchorEl={anchorElNav}
            anchorOrigin={{
              vertical: 'bottom',
              horizontal: 'left',
            }}
            keepMounted
            transformOrigin={{
              vertical: 'top',
              horizontal: 'left',
            }}
            open={Boolean(anchorElNav)}
            onClose={handleCloseNavMenu}
            onClick={handleCloseNavMenu}
            sx={{
              display: { xs: 'block', md: 'none' },
            }}
          >
            <MenuItem component={Link} to="/" onClick={handleCloseNavMenu}>
              <ListItemIcon>
                <ArrowRight fontSize="small" />
              </ListItemIcon>
              Inicio
            </MenuItem>
            <MenuItem component={Link} to="/create" onClick={handleCloseNavMenu}>
              <ListItemIcon>
                <ArrowRight fontSize="small" />
              </ListItemIcon>
              Agregar receta
            </MenuItem>
          </Menu>
        </Box>

        <Typography
          variant="h5"
          noWrap
          component="a"
          href=""
          sx={{
            mr: 3,
            mb: '5px',
            display: { xs: 'flex', md: 'none' },
            flexGrow: 1,
            fontFamily: 'monospace',
            fontWeight: 700,
            letterSpacing: '.3rem',
            color: 'inherit',
            textDecoration: 'none',
          }}
        >
          Recipes App
        </Typography>

        <Box sx={{ flexGrow: 1, display: { xs: 'none', md: 'flex' } }}>
          <Button component={Link} to="/" onClick={handleCloseNavMenu} sx={{ my: 2, color: 'white', display: 'block' }} >
            Inicio
          </Button>
          <Button component={Link} to="/create" onClick={handleCloseNavMenu} sx={{ my: 2, color: 'white', display: 'block' }} >
            Agregar receta
          </Button>
        </Box>

        <Box sx={{ flexGrow: 0 }}>
          <Chip sx={{ display: { xs: 'none', md: 'inline-flex' } }} label={<span style={{ color: '#fff' }}>Bienvenido {user.email}</span>} />
          <Button color="error" size="small" onClick={onLogout}>
            <ExitToAppOutlined />
            Salir
          </Button>
        </Box>

      </Toolbar>
    </AppBar>
  )
}
