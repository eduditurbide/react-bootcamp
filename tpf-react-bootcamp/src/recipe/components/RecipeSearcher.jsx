import { Clear, Search } from '@mui/icons-material'
import { Grid, IconButton, InputAdornment, InputBase, TextField } from '@mui/material'
import React, { useRef } from 'react'
import { useForm } from '../../hooks'

export const RecipeSearcher = ({ onSearchEvent = null}) => {

  const searchInputRef = useRef()

  const { 
    toSearch, 
    onInputChange,  
    onResetForm,
  } = useForm({
    toSearch: ''
  })

  const onSearch = (e) => {
    e?.preventDefault()
    onSearchInRecipeList()
  }

  const onHandleKeyPress = ({ key, keyCode }) => {
    if (key === 'Enter' || keyCode === 13) {
      onSearchInRecipeList()
    }
  }
  
  const onSearchInRecipeList = () => {
    if (toSearch?.trim().length <= 1) { return; }
    onSearchEvent(toSearch)
    searchInputRef.current.focus()
  }

  const onClear = () => { 
    onResetForm()
    onSearchEvent(null)
    searchInputRef.current.focus()
  }

  return (
    <Grid container alignItems={"center"} m={1}>
      <InputBase 
        inputRef={searchInputRef}
        variant="standard"
        sx={{ ml: 1, flex: 1 }}
        placeholder="Buscador de recetas"
        inputProps={{ 'aria-label': 'Buscador de recetas.' }}
        name='toSearch'
        value={toSearch}
        onChange={onInputChange}
        onKeyUp={onHandleKeyPress}
        endAdornment={ toSearch && (
          <InputAdornment position="end">
            <IconButton
              aria-label="Eliminar filtro de bÃºsqueda."
              onClick={onClear}
              edge="end"
            >
              <Clear color="neutral" />
            </IconButton>
          </InputAdornment>
        )}
      />
      <IconButton 
        type="button" 
        sx={{ p: '10px' }} 
        aria-label="search"
        onClick={onSearch}
      >
        <Search />
      </IconButton>
    </Grid>
  )
}
