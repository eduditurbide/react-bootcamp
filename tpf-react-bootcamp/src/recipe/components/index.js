export * from './DeleteButton'
export * from './Navbar'
export * from './RecipeForm'
export * from './RecipeList'
export * from './RecipeSearcher'