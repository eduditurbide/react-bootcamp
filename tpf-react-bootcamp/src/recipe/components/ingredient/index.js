export * from './IngredientAdd'
export * from './IngredientForm'
export * from './IngredientList'