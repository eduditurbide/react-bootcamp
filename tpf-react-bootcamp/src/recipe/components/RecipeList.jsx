import { InsertPhotoOutlined, Search, TurnedInNot } from "@mui/icons-material"
import { Backdrop, Box, CircularProgress, Divider, Grid, IconButton, InputBase, List, ListItem, ListItemButton, ListItemIcon, ListItemText, TextField, Toolbar, Typography } from "@mui/material"
import { useRecipes } from "../hooks/useRecipes"
import { useEffect } from "react"
import { RecipeSearcher } from "./RecipeSearcher"

export const RecipeList = () => {

  const { recipeList, selected, onSelectRecipe, getRecipesList, isLoading } = useRecipes()

  useEffect(() => {
    getRecipesList()
  }, [])
  
  const onSearchRecipe = (word) => { 
    getRecipesList(word)
  }

  return (
    <Box
      component='nav'
      sx={{ flexShrink: { sm: 0 } }}
    >
      <Toolbar>
        <Typography variant="h6" noWrap component={'div'}>
          Listado de recetas
        </Typography>
      </Toolbar>

      <Divider />  
      <RecipeSearcher onSearchEvent={e => onSearchRecipe(e)}/>
      <Divider sx={{ mb: 1 }}/>  

      <List>
        {(Array.isArray(recipeList) && recipeList.length > 0) && (recipeList.map((recipe) => (
          <ListItem 
            key={recipe._id} 
            disablePadding
            secondaryAction={
              <IconButton 
                aria-label="Marcar receta como favorita." 
                edge="end" 
                size="large"
              >
                <TurnedInNot />
              </IconButton>
            }
          >
            <ListItemButton 
              sx={{ gap: 3 }}
              onClick={e => onSelectRecipe(selected?._id === recipe._id ? false : recipe._id)}
              selected={selected?._id === recipe._id}
            >
              {recipe?.imagePath && (
                <img src={recipe.imagePath} alt="Img de receta" style={{ height: '45px', width: '45px', objectFit: 'cover' }}/>
              )}
              {!recipe?.imagePath && (
                <InsertPhotoOutlined color="disabled" sx={{ fontSize: 50 }} />
              )}
              <Grid container>
                <ListItemText primary={recipe.name} />
              </Grid>
            </ListItemButton>
          </ListItem>
        )))
        }
      </List>

      <Backdrop
        sx={{ color: '#fff', zIndex: (theme) => theme.zIndex.drawer + 1 }}
        open={isLoading}
      >
        <CircularProgress color="inherit" />
      </Backdrop>
    </Box>
  )
}
