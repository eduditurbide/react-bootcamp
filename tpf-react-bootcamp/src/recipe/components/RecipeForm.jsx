import { Button, Grid, TextField } from "@mui/material"
import { useForm } from "../../hooks"
import { SaveOutlined } from "@mui/icons-material"
import { IngredientForm } from "./ingredient"
import { DeleteButton } from "./DeleteButton"

const initRecipeForm = {
  name: "",
  description: "",
  ingredients: [],
  imagePath: "",
}

export const RecipeForm = ({recipe = null, onSubmit = null, onDelete = null}) => {

  const { 
    formState,
    name, 
    description, 
    ingredients, 
    imagePath, 
    onInputChange,
  } = useForm(recipe || initRecipeForm) 

  const onSaveRecipe = () => {
    if (onSubmit) {
      onSubmit(formState)
    }
  }

  return (
    <>
      <Grid container gap={3}>
        <TextField
          type="text"
          variant="filled"
          fullWidth
          sx={{ border: 'none' }}
          placeholder="Ingrese un título"
          label="Título"
          name="name"
          value={name}
          onChange={onInputChange}
        />

        <TextField
          type="url"
          variant="filled"
          fullWidth
          sx={{ border: 'none' }}
          placeholder="Ingrese el link de una imagen"
          label="Imagen"
          name="imagePath"
          value={imagePath}
          onChange={onInputChange}
        />

        {imagePath && ( <img alt="Imagen de receta." src={imagePath} style={{ height: '200px', width: 'auto', mb: 3 }}/> )}

        <TextField
          type="text"
          variant="filled"
          fullWidth
          sx={{ border: 'none' }}
          multiline
          placeholder="Describe el tipo de comida y los pasos a seguir para preparala ;)"
          label="Descripción"
          name="description"
          value={description}
          onChange={onInputChange}
        />

        <IngredientForm name="ingredients" ingredientList={ingredients} onIngredientsChange={onInputChange}/>
      
        <Grid item xs={12}>
          <Grid container alignItems={"center"} justifyContent={"center"} gap={2}>
            <Button 
              variant="contained" 
              color="primary" 
              sx={{ padding: 1, width: { xs: '100%', sm: 'auto' } }} 
              onClick={onSaveRecipe}
            >
              <SaveOutlined sx={{ fontSize: 30, mr: 1 }} />
              Guardar
            </Button>

            {recipe?._id && (
              <DeleteButton
                title="Eliminar receta"
                text={`Estas por eliminar la receta "${recipe.name}", ¿Deseas continuar?`}
                onAgreeClickEvent={e => onDelete ? onDelete(formState) : null}
              />
            )}
          </Grid>
        </Grid>

      </Grid>
    </>
  )
}
