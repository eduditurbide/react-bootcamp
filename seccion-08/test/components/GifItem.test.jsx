import {render, screen} from '@testing-library/react'
import { GifItem } from '../../src/components/GifItem'

describe('Test de <GitItem/>', () => { 

  const title = 'Saitama'
  const url = 'https://one-punch.com/saitama.jpg'

  /** Test 1 */
  test('debe de hacer match con el snaptshot', () => { 
    const {container} = render(<GifItem title={title} url={url} />)
    expect(container).toMatchSnapshot()
  })

  /** Test 2 */
  test('debe de mostrar la imagen con el URL y el ALT indicado', () => {

    render(<GifItem title={title} url={url} />)
    // screen.debug()
    const {src, alt} = screen.getByRole('img')
    expect(src).toBe(url)
    expect(alt).toBe(title)

  })

  /** Test 3 */
  test('debe de mostrar el titulo en el componente', () => {

    render(<GifItem title={title} url={url} />)
    expect(screen.getByText(title)).toBeTruthy()

  })

})