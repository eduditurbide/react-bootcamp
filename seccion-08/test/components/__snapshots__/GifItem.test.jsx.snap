// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test de <GitItem/> debe de hacer match con el snaptshot 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="Saitama"
      src="https://one-punch.com/saitama.jpg"
    />
    <p>
      Saitama
    </p>
  </div>
</div>
`;
